### PR TITLE
Support OpenTracing 0.32.0 and 0.33.0

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

**Comment by pavolloffay** [comment]
wait tracerresolver has to be updated too

File: jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java [hunk 1]
```diff
@@ -475,23 +475,12 @@ public JaegerSpan start() {
       return jaegerSpan;
     }
 
-    @Override
-    public Scope startActive(boolean finishSpanOnClose) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: Do we really want to do this? I think the idea of 0.33 was to give users a cleaner API that will facilitate transition to OpenTelemetry, but implementations were still advised to support 0.32 for better backwards compatibility.

cc @carlosalberto

Comment (carlosalberto) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: Well, we won't even support any of the deprecated members (we throw an exception from these methods in the OT shim), so it's not a big difference.

Buuut if you think API or deployment wise we should nevertheless reference 0.32, then lets talk about it and consider reverting the latest update for this in the OT shim.

Comment (yurishkuro) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: I think I did not explain myself. Imagine an application X instrumented with 0.32 or even 0.31 API. If we do this PR to jaeger client, then  that application X will no longer be able to upgrade to newer versions of Jaeger client, even though technically there are no major breaking changes. At the same time, another application Y instrumented with 0.33 API would still be able to use Jaeger client compiled against 0.32.

Thus we can maintain broader compatibility of Jaeger client.

Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: What are your proposals? Keep the client on 0.32.0 as it is a superset of 0.33.0?

**Comment by bltavares** [comment]
I've recently found the error where the library is instrumented using `opentracing 0.33` (on [io.pedestal.log](https://github.com/pedestal/pedestal/blob/master/log/project.clj#L25)) could not use the `0.35` version of Jaeger.

It failed at runtime, when `JaegerTracer` tried to call `.activeSpan` on a `scopeManager` [[here](https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java#L194)]

<details>
<summary>Stack trace</summary>

```
:via
              [{:type java.lang.NoSuchMethodError
                :message "io.opentracing.ScopeManager.activeSpan()Lio/opentracing/Span;"
                :at [io.jaegertracing.internal.JaegerTracer activeSpan "JaegerTracer.java" 194]}]
              :trace
              [[io.jaegertracing.internal.JaegerTracer activeSpan "JaegerTracer.java" 194]
               [io.opentracing.util.GlobalTracer activeSpan "GlobalTracer.java" 148]
               [io.pedestal.log$eval33020$fn__33030 invoke "log.clj" 845]
```
</details>

This class has been modified, but the method haven't. Could the error have been caused by version conflicts of the interface on the classpath, or is this something should be tackled by the bump as well?

(I didn't have time yet to test it with this PR code tho, sorry about that)

**Comment by objectiser** [comment]
@bltavares Just tried calling `activeSpan()` via the `GlobalTracer`, using Jaeger 0.35.2 and OpenTracing 0.33, and it worked fine - could you double check the jaeger and OT dependencies being used?

**Comment by bltavares** [comment]
@objectiser It seems that it is loading another version of `opentracing` from pedestal. My bad, I've looked only on GitHub for the source of the lib, using the latest code instead of the version.

It fails with Opentracing `0.31.0` and Jaeger `0.35.5`. I'll post the dep tree for sake of completeness.

<details>
<summary>dep tree</summary>

```
+- io.pedestal:pedestal.log:jar:0.5.5:compile
[INFO] |  |  +- (org.clojure:clojure:jar:1.9.0:compile - omitted for conflict with 1.10.0)
[INFO] |  |  +- (org.slf4j:slf4j-api:jar:1.7.25:compile - omitted for duplicate)
[INFO] |  |  +- (io.dropwizard.metrics:metrics-core:jar:4.0.2:compile - omitted for conflict with 3.2.3)
[INFO] |  |  +- io.dropwizard.metrics:metrics-jmx:jar:4.0.2:compile
[INFO] |  |  |  +- (io.dropwizard.metrics:metrics-core:jar:4.0.2:compile - omitted for conflict with 3.2.3)
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.25:compile - omitted for duplicate)
[INFO] |  |  +- io.opentracing:opentracing-api:jar:0.31.0:compile
[INFO] |  |  \- io.opentracing:opentracing-util:jar:0.31.0:compile
[INFO] |  |     +- (io.opentracing:opentracing-api:jar:0.31.0:compile - omitted for duplicate)
[INFO] |  |     \- io.opentracing:opentracing-noop:jar:0.31.0:compile
[INFO] |  |        \- (io.opentracing:opentracing-api:jar:0.31.0:compile - omitted for duplicate)
[INFO] |  +- io.jaegertracing:jaeger-client:jar:0.35.5:compile
[INFO] |  |  +- io.jaegertracing:jaeger-thrift:jar:0.35.5:compile
[INFO] |  |  |  +- (io.jaegertracing:jaeger-core:jar:0.35.5:compile - omitted for duplicate)
[INFO] |  |  |  +- (org.slf4j:slf4j-api:jar:1.7.25:compile - omitted for duplicate)
[INFO] |  |  |  +- (org.apache.thrift:libthrift:jar:0.12.0:compile - omitted for conflict with 0.10.0)
[INFO] |  |  |  \- com.squareup.okhttp3:okhttp:jar:3.9.0:compile
[INFO] |  |  |     \- com.squareup.okio:okio:jar:1.13.0:compile
[INFO] |  |  +- io.jaegertracing:jaeger-core:jar:0.35.5:compile
[INFO] |  |  |  +- (io.opentracing:opentracing-api:jar:0.32.0:compile - omitted for conflict with 0.31.0)
[INFO] |  |  |  +- (io.opentracing:opentracing-util:jar:0.32.0:compile - omitted for conflict with 0.31.0)
[INFO] |  |  |  +- (com.google.code.gson:gson:jar:2.8.2:compile - omitted for conflict with 2.3.1)
[INFO] |  |  |  \- (org.slf4j:slf4j-api:jar:1.7.25:compile - omitted for duplicate)
[INFO] |  |  \- io.jaegertracing:jaeger-tracerresolver:jar:0.35.5:compile
[INFO] |  |     +- (io.jaegertracing:jaeger-core:jar:0.35.5:compile - omitted for duplicate)
[INFO] |  |     \- io.opentracing.contrib:opentracing-tracerresolver:jar:0.1.6:compile
[INFO] |  |        \- (io.opentracing:opentracing-api:jar:0.32.0:compile - omitted for conflict with 0.31.0)
````
</details>

File: jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java (hunk 1 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: I have added the removed methods back. To make sure it works I will add a test submodule for 0.32.0 to assure it works.

File: jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java [hunk 2]
```diff
@@ -475,9 +475,20 @@ public JaegerSpan start() {
...
+    @Deprecated
+    // @Override keep compatibility with 0.32.0
+    public Scope startActive(final boolean finishSpanOnClose) {
+      final Span span = start();
+      final Scope wrapped = scopeManager.activate(span);
+      return new Scope() {
+        @Override
+        public void close() {
+          wrapped.close();
+          if (finishSpanOnClose) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: nit: could short-circuit this at the top
```
if (finishSpanOnClose) {
    return scopeManager.activate(start());
}
```

File: build.gradle [hunk 3]
```diff
@@ -17,13 +17,13 @@ plugins {
...
 
-ext.opentracingVersion = getProperty('opentracingVersion','0.32.0')
+ext.opentracingVersion = getProperty('opentracingVersion','0.33.0')
 ext.guavaVersion = getProperty('guavaVersion','18.0')
 ext.apacheThriftVersion = getProperty('apacheThriftVersion','0.12.0')
 ext.jerseyVersion = getProperty('jerseyVersion','2.22.2')
 ext.slf4jVersion = getProperty('slf4jVersion','1.7.25')
 ext.gsonVersion = getProperty('gsonVersion','2.8.2')
-ext.tracerResolverVersion = getProperty('tracerResolverVersion','0.1.6')
+ext.tracerResolverVersion = getProperty('tracerResolverVersion','0.1.7')
```
Comment (objectiser) [inline build.gradle]: Version 0.1.8 has been released with a couple of changes (adding module name and supporting classloader parameter).

File: jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java [hunk 4]
```diff
@@ -486,9 +503,9 @@ public Scope startActive(boolean finishSpanOnClose) {
       return this;
     }
 
-    @Override
     @Deprecated
-    public JaegerSpan startManual() {
+    // @Override keep compatibility with 0.32.0
+    public Span startManual() {
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: No, It has to be `Span` now as it does not override the API anymore.

File: jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java [hunk 5]
```diff
@@ -475,9 +475,26 @@ public JaegerSpan start() {
...
+        Span span = start();
+        Scope wrapped = scopeManager.activate(span);
+        @Override
+        public void close() {
+          wrapped.close();
+            span.finish();
+        }
+
+        // @Override keep compatibility with 0.32.0
+        public Span span() {
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/internal/JaegerTracer.java]: See the previous message, we cannot return covariant type as the method does not override.

File: jaeger-core/src/test/java/io/jaegertracing/internal/ActiveSpanTest.java [hunk 6]
```diff
@@ -89,25 +94,30 @@ public void testActiveSpanAutoReference() {
   }
 
   @Test
+  @SuppressWarnings("deprecation")
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply jaeger-core/src/test/java/io/jaegertracing/internal/ActiveSpanTest.java]: I am fine keeping it here too. Here it uses the direct type whereas in the 0.32 test it uses old interface.

**Pull request merged**